Fix cleanenv

ArangoGutierrez opened the pull request:
**Description of the Pull Request (PR):**



**This fixes or addresses the following GitHub issues:**

- Ref: 
--cleanenv seems to be making the environment TOO clean #1893


**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [ ] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [x] I have tested this PR locally with a `make test`
- [x] This PR is against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge


Attn: @singularityware-admin

ArangoGutierrez committed:
Create a CleanEnv func under util pkg

Func CleanEnv cleans environment variables before running the container
it recieves  a pointer to the generator, the host environment and
the --cleanenv flag value.

ArangoGutierrez committed:
Use util func CleanEnv

import src/pkg/util/env/ and use func CleanEnv() to set the ENV vars
to be set inside the container runtime

jmstover commented on src/pkg/util/env/clean.go:
```diff
@@ -0,0 +1,41 @@
+	for _, env := range environment {
+		e := strings.SplitN(env, "=", 2)
+		if len(e) != 2 {
+			sylog.Verbosef("can't process environment variable %s", env)
```
Capitalize `Can't` here...

ArangoGutierrez committed:
Capitalize can't on CleanEnv

on src/pkg/util/env/clean/go, at line 21, capitalize the first word of the
verbose output

jmstover reviewed (dismissed):
This works. Only difference between this and 2.x is the 3.0 version doesn't set `LANG=C`, and also the variables `SINGULARITY_CONTAINER` and `SINGULARITY_NAME` aren't set.

yhcote committed:
Merge branch 'master' into fix_cleanenv

GodloveD commented on src/pkg/util/env/clean.go:
```diff
@@ -0,0 +1,41 @@
+			e[0] != "no_proxy" &&
+			e[0] != "all_proxy" &&
+			e[0] != "ftp_proxy") {
+			continue
```
I would include the SINGULARITY_NAME and SINGULARITY_CONTAINER variables here.  And then we must set them somewhere before this code executes.

bauerm97 commented on src/cmd/singularity/cli/actions.go:
```diff
@@ -269,30 +270,7 @@ func execStarter(cobraCmd *cobra.Command, image string, args []string, name stri
-			continue
-		}
-
-		if e[0] == "HOME" {
```
How does the home env variable get set if you're removing the handling from here?

bauerm97 commented on src/pkg/util/env/clean.go:
```diff
@@ -0,0 +1,41 @@
+
+// CleanEnv cleans environment variables before running the container
+func CleanEnv(g *generate.Generator, IsCleanEnv bool, environment []string) {
+
```
I would prefer if we don't leave lines of space here

ArangoGutierrez replied:
umm will check my VScode config on this one

bauerm97 commented on src/pkg/util/env/clean.go:
```diff
@@ -0,0 +1,41 @@
+		// Transpose environment
+		if strings.HasPrefix(e[0], "SINGULARITYENV_") {
+			e[0] = strings.TrimPrefix(e[0], "SINGULARITYENV_")
+		} else if IsCleanEnv && (e[0] != "HOME" &&
```
This seems like we're doing things in a very naive way here. First of all, we're not guaranteed that these variables are set yet. For instance in 2.x, we always force the env var `LANG=C` to be set (however, this was to fix some bugs and I *don't* think that needs to happen in 3.0)

GodloveD replied:
I spoke with Greg about the `LANG=C` thing and I think he just set it b/c it is critical for many apps (like Perl) and he was tired of getting issues raised about things being broken because of it.  IMHO we should leave it because we have already gone back and forth about it once.  It's a judgement call, a judgement has been made, and now it's important to just be consistent.

ArangoGutierrez replied:
so... final call : 
 - force `LANG=C`?
 - or just pass the `LANG` env value from the host?

bauerm97 replied:
Always set `LANG=C` I guess

ArangoGutierrez replied:
From 2.6 
```
$ singularity exec ~/tmp/sing/centos7.img env | grep ^LANG
LANG=en_US
$ singularity exec --cleanenv ~/tmp/sing/centos7.img env | grep ^LANG 
LANG=C
```
so force `LANG=C` is just when using `--cleanenv` flag

bauerm97 commented on src/pkg/util/env/clean.go:
```diff
@@ -0,0 +1,41 @@
+		}
+
+		// Transpose environment
+		if strings.HasPrefix(e[0], "SINGULARITYENV_") {
```
Also, since this is in an if...else if block, this implies that `SINGULARITYENV_` variables are *always* transformed even when `--cleanenv` is set, is that what we want?

GodloveD replied:
Yes that is the behavior in 2.6.

ArangoGutierrez committed:
Remove white line on clean.go

ArangoGutierrez committed:
Pass --no-home flag value to CleanEnv()

ArangoGutierrez committed:
Add SINGULARITY_CONTAINER and SINGULARITY_NAME env vars

At Transpose environment step, add 2 new env vars  SINGULARITY_CONTAINER
and SINGULARITY_NAME

ArangoGutierrez committed:
Merge branch 'master' into fix_cleanenv

ArangoGutierrez committed:
Merge branch 'fix_cleanenv' of github.com:ArangoGutierrez/singularity into fix_cleanenv

ArangoGutierrez committed:
better format on clean.go

bauerm97 commented on src/pkg/util/env/clean.go:
```diff
@@ -0,0 +1,50 @@
+			continue
+		}
+
+		if e[0] == "HOME" {
```
This conditional doesn't make a lot of sense to me. I think we should always set `HOME=HomeDest` inside the container, whether or not the `HOME` variable is already set on this host. This conditional implies that we're only setting `HOME` in the container if `HOME` is *already* set on the host.

In my opinion, it makes much more sense to set the unconditional env variables outside of the loop where we iterate over the environment:

```
for ... {
    // iterating over env
}

if !NoHome {
    g.AddProcessEnv("HOME", HomeDest)
} else {
    g.AddProcessEnv("HOME", "/")
}

g.AddProcessEnv("LANG", "C")
```

jmstover replied:
That's opposite current behavior:
```
$ /usr/local/singularity/2.6.0/bin/singularity shell -c --no-home ~/tmp/sing/centos7.img 
WARNING: Could not chdir to home: /home/jason
Singularity centos7.img:/> pwd
/
Singularity centos7.img:/> echo $HOME
/home/jason
```

The above is the same behavior as if you SSH into a node, but your `$HOME` isn't availible. You get sent to `/` as your CWD, but `$HOME` is still set to what your HOME actually is.

What you are saying is when `--no-home` is specified, then `HOME=/` and that's just wrong from a system standpoint.

ArangoGutierrez replied:
```
singularity exec --help 
[...] 
      --no-home            Do NOT mount users home directory if home is
                           not the current working directory.
```
so @jmstover is right, `--no-home` should just NOT mount HOME

bauerm97 commented on src/pkg/util/env/clean.go:
```diff
@@ -0,0 +1,50 @@
+)
+
+// CleanEnv cleans environment variables before running the container
+func CleanEnv(g *generate.Generator, NoHome bool, IsCleanEnv bool, HomeDest string, environment []string) {
```
I think this func should be called something like `SetContainerEnv()` instead of `CleanEnv()` to avoid confusion with the `--cleanenv` flag

bauerm97 commented on src/pkg/util/env/clean.go:
```diff
@@ -0,0 +1,50 @@
+			e[0] != "no_proxy" &&
+			e[0] != "all_proxy" &&
+			e[0] != "ftp_proxy" &&
+			e[0] != "SINGULARITY_CONTAINER" &&
```
These two variables still aren't ever being set (`SINGULARITY_CONTAINER` && `SINGULARITY_NAME`). From what I remember, `SINGULARITY_CONTAINER` is image argument passed to the CLI and `SINGULARITY_NAME` is the name of the file. For 
example:

```
singularity run /my/test/image.sif

SINGULARITY_CONTAINER=/my/test/image.sif
SINGULARITY_NAME=image.sif
```

ArangoGutierrez replied:
and when the image comes from a URI?
e.g
```
singularity run library://alpine:latest
```

ArangoGutierrez replied:
should we leave `SINGULARITY_NAME` when actually having a name, a.k.a when using instances?

jmstover replied:
@bauerm97 ... look at `/.singularity.d/actions/shell` .... `$SINGULARITY_CONTAINER` is being used in that to set PS1 when bash exists as a shell in the container.

Here's a comparison of 2.6.0 and 3.0.0 on the same image:
```
$ /usr/local/singularity/2.6.0/bin/singularity shell ~/tmp/sing/centos7.img 
Singularity centos7.img:~/development/go/src/github.com/singularityware/singularity

$ /usr/local/singularity/3.0.0-dev/bin/singularity shell ~/tmp/sing/centos7.img
Singularity :~/development/go/src/github.com/singularityware/singularity>
```

jmstover replied:
@ArangoGutierrez 

```
$ /usr/local/singularity/2.6.0/bin/singularity shell docker://centos:7
  [ ... ]
SINGULARITY_NAME=centos:7
SINGULARITY_CONTAINER=centos:7
```

ArangoGutierrez replied:
they *Should* be different , at least `CONTAINER` should point the path/source, i kind of agree with the name. I would use `NAME` when running instances. but is just my opinion

ArangoGutierrez replied:
```
[eduardo@rhel7 singularity]$ singularity exec /tmp/alpine.sif env | grep SINGULARITY
SINGULARITY_CONTAINER=/tmp/alpine.sif
SINGULARITY_NAME=alpine.sif
```
and with instances
```
[eduardo@rhel7 singularity]$ singularity instance start image.sif namy
[eduardo@rhel7 singularity]$ singularity exec instance://namy env | grep SINGULARITY
SINGULARITY_CONTAINER=instance://namy
SINGULARITY_NAME=namy
```

@GodloveD , @bauerm97  LMK if this is the right behavior so I can go ahead and push the changes

ArangoGutierrez committed:
Set SINGULARITY_* CONTAINER and NAME vars properly

SINGULARITY_NAME = image file name
SINGULARITY_CONTAINER = full path to image

ArangoGutierrez committed:
rename CleanEnv to SetContainerEnv

ArangoGutierrez committed:
Force LANG env only when using --cleanenv flag

ArangoGutierrez committed:
Merge branch 'master' into fix_cleanenv

ArangoGutierrez committed:
Merge branch 'fix_cleanenv' of github.com:ArangoGutierrez/singularity into fix_cleanenv

ArangoGutierrez committed:
Merge branch 'master' into fix_cleanenv

ArangoGutierrez committed:
Use SINGULARITY_NAME on shellFileContent

on scr/pkg/build/base_environment
change SINGULARITY_CONTAINER for SINGULARITY_NAME on shellFileContent const

bauerm97 committed:
Clean up environment handling function

ArangoGutierrez committed:
Merge pull request #4 from bauerm97/fix-cleanenv

Fix-cleanenv cleanup

ArangoGutierrez committed:
Merge branch 'master' into fix_cleanenv

GodloveD reviewed (approved):
lgtm

ArangoGutierrez committed:
Merge branch 'master' into fix_cleanenv
